### PR TITLE
feat: add ML filter resource

### PR DESCRIPTION
**Pull request by edsavage**
## Summary

- Add `elasticstack_elasticsearch_ml_filter` resource for managing ML filters used in anomaly detection job `custom_rules`

## Changelog

Customer impact: enhancement
Summary: Add `elasticstack_elasticsearch_ml_filter` resource for managing Elasticsearch ML filters (used with anomaly detection `custom_rules`).

- Adds a first-class Terraform resource for Elasticsearch ML filters so practitioners can manage filter contents and lifecycle in code (including filters referenced from anomaly detection `custom_rules`) instead of relying on Kibana-only workflows or one-off API calls. ([#1970](https://github.com/elastic/terraform-provider-elasticstack/pull/1970))

## Details

Supports full CRUD with in-place updates via the dedicated Update filter API. `description` and `items` (set of strings) can both be updated in-place — items diffs are computed and sent as `add_items`/`remove_items`. `filter_id` requires replacement.

## Test plan

- [x] Unit tests: 8 cases covering model conversions, null/empty handling
- [x] Acceptance tests: 3 cases (create+update lifecycle, no-items edge case, import)
- [x] `make build` passes (including doc generation)

Made with [Cursor](https://cursor.com)

> [!NOTE]
> ### Add `elasticstack_elasticsearch_ml_filter` resource for Elasticsearch ML filters
> - Implements full CRUD lifecycle for ML filters via `PUT /_ml/filters/{filter_id}`, `GET /_ml/filters/{filter_id}`, `POST /_ml/filters/{filter_id}/_update`, and `DELETE /_ml/filters/{filter_id}`.
> - Update diffs current vs. planned items to compute `add_items`/`remove_items` for the update API.
> - Schema validates `filter_id` with length (1–64) and regex constraints; changing `filter_id` forces replacement.
> - Registers the new resource in the provider via [plugin_framework.go](https://github.com/elastic/terraform-provider-elasticstack/pull/1970/files#diff-5a66d72c3748bd201add43c9271e7b8dc29f28fd06eee792c837aa87e70a063b) and includes acceptance and unit tests.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 5c80f89. (Automatic summaries will resume when PR exits draft mode or review begins).</sup>
> 
>
>

### CODE REVIEW

File: internal/elasticsearch/ml/filter/create.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		return plan, diags
+	}
+
+	body, err := json.Marshal(apiModel)
+	if err != nil {
+		diags.AddError("Failed to marshal filter configuration", err.Error())
+		return plan, diags
+	}
+
+	_, err = typedClient.Ml.PutFilter(filterID).Raw(bytes.NewReader(body)).Do(ctx)
```
Comment (tobio) [inline internal/elasticsearch/ml/filter/create.go]:
```suggestion
	_, err = typedClient.Ml.PutFilter(filterID).
		Description(plan.Description.ValueString()).
		Items(items...).Do(ctx)
```

Any reason we can't do something like this? Generally, do we need to be redefining the API model here anymore?

File: internal/elasticsearch/ml/filter/read.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	diags.Append((&out).fromAPIModel(ctx, api)...)
+	if diags.HasError() {
+		return state, false, diags
+	}
+
+	tflog.Debug(ctx, fmt.Sprintf("Successfully read ML filter: %s", filterID))
+	return out, true, diags
+}
+
+func mlFilterToAPIModel(f *types.MLFilter) *APIModel {
```
Comment (tobio) [inline internal/elasticsearch/ml/filter/read.go]: Surely we can get rid of this and have `fromAPIModel` consume `*types.MLFilter` directly.

File: internal/elasticsearch/ml/filter/schema.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+				},
+			},
+			"description": schema.StringAttribute{
+				MarkdownDescription: "A description of the filter.",
+				Optional:            true,
+			},
+			"items": schema.SetAttribute{
+				MarkdownDescription: "The items of the filter. A wildcard `*` can be used at the beginning or the end of an item. Up to 10000 items are allowed in each filter.",
+				Optional:            true,
+				ElementType:         types.StringType,
```
Comment (tobio) [inline internal/elasticsearch/ml/filter/schema.go:71]:
```suggestion
				ElementType:         types.StringType,
				Validators: []validator.Set{
					setvalidator.SizeAtMost(10000),
				},
```

Seems a bit wild to think someone will hit this, but since we explicitly call it out.

File: internal/elasticsearch/ml/filter/delete.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	tflog.Debug(ctx, fmt.Sprintf("Deleting ML filter: %s", filterID))
+
+	typedClient, err := client.GetESClient()
+	if err != nil {
+		diags.AddError("Failed to get Elasticsearch client", err.Error())
+		return diags
+	}
+
+	_, err = typedClient.Ml.DeleteFilter(filterID).Do(ctx)
+	if err != nil {
```
Comment (tobio) [inline internal/elasticsearch/ml/filter/delete.go:49]: At the moment this will fail if the filter has been deleted outside of Terraform. We should handle 404s here and ignore them. 

```suggestion
	if err != nil && !clients.IsNotFoundElasticsearch(err) {
```

File: internal/elasticsearch/ml/filter/models.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+	return apiModel, diags
+}
+
+func (m *TFModel) fromAPIModel(ctx context.Context, apiModel *APIModel) fwdiags.Diagnostics {
+	var diags fwdiags.Diagnostics
+
+	m.FilterID = types.StringValue(apiModel.FilterID)
+
+	if apiModel.Description != "" {
```
Comment (tobio) [inline internal/elasticsearch/ml/filter/models.go]: We treat `""` as null here, but not on create/update. As it stands `description = ""` will likely have a perpetual diff or a state consistency error. 

We could just validate a non-zero length description in the schema if that makes sense. If not we need to be more explicit here.

**Review by tobio — COMMENTED** [review]
Pull the anomaly detection scope changes out to another PR please.

File: internal/elasticsearch/ml/filter/resource.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+func (r *filterResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_elasticsearch_ml_filter"
+}
+
+func (r *filterResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	client, diags := clients.ConvertProviderData(req.ProviderData)
+	resp.Diagnostics.Append(diags...)
+	r.client = client
+}
```
[the remark this answers is not in the document]
Comment (edsavage) [reply internal/elasticsearch/ml/filter/resource.go:58]: **Resolved** — This predates the entitycore migration. The resource now uses `entitycore.NewElasticsearchResource` with `ProviderClientFactory`, scoped Elasticsearch clients, and `GetElasticsearchClient(plan/elasticsearch_connection)` so overrides work (see `internal/entitycore/resource_envelope.go`).

File: internal/elasticsearch/ml/filter/resource.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+func (r *filterResource) resourceReady(diags *fwdiags.Diagnostics) bool {
+	if r.client == nil {
+		diags.AddError("Client not configured", "Provider client is not configured")
+		return false
+	}
+	return true
+}
+
+func (r *filterResource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
```
[the remark this answers is not in the document]
Comment (edsavage) [reply internal/elasticsearch/ml/filter/resource.go]: **Resolved** — `ImportState` parses the composite id and sets both `id` and `filter_id` via `clients.CompositeIDFromStrFw` / `compID.ResourceID`, matching other ML resources.

File: internal/elasticsearch/ml/filter/schema.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	resp.Schema = GetSchema()
+}
+
+func GetSchema() schema.Schema {
+	return schema.Schema{
+		MarkdownDescription: "Manages Machine Learning filters. A filter contains a list of strings that can be referenced " +
+			"in the `custom_rules` property of detector configuration objects. " +
+			"See the [ML Filter API documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/ml-put-filter.html) for more details.",
+		Blocks: map[string]schema.Block{
+			"elasticsearch_connection": providerschema.GetEsFWConnectionBlock(false),
```
[the remark this answers is not in the document]
Comment (edsavage) [reply internal/elasticsearch/ml/filter/schema.go]: **Resolved** — The connection block is injected by `ElasticsearchResource.Schema` (`providerschema.GetEsFWConnectionBlock()` with no args). This package's `getSchema` does not call `GetEsFWConnectionBlock(false).

File: internal/elasticsearch/ml/filter/create.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tflog.Debug(ctx, fmt.Sprintf("Creating ML filter: %s", filterID))
+
+	esClient, err := r.client.GetESClient()
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to get Elasticsearch client", err.Error())
+		return
```
[the remark this answers is not in the document]
Comment (edsavage) [reply internal/elasticsearch/ml/filter/create.go]: **Resolved** — Create runs through the envelope callback with `*ElasticsearchScopedClient`; `createFilter` uses `client.GetESClient()` and `client.ID(...)` on that scoped client.

File: internal/elasticsearch/ml/filter/delete.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	var filterIDValue basetypes.StringValue
+	diags := req.State.GetAttribute(ctx, path.Root("filter_id"), &filterIDValue)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	filterID := filterIDValue.ValueString()
+
+	tflog.Debug(ctx, fmt.Sprintf("Deleting ML filter: %s", filterID))
```
[the remark this answers is not in the document]
Comment (edsavage) [reply internal/elasticsearch/ml/filter/delete.go]: **Resolved** — Delete receives `resourceID` from `entitycore.Delete`, which parses the composite `model.GetID()` and passes `compID.ResourceID` — not `filter_id` alone.

File: internal/elasticsearch/ml/filter/update.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+	compID, sdkDiags := clients.CompositeIDFromStr(state.ID.ValueString())
+	resp.Diagnostics.Append(diagutil.FrameworkDiagsFromSDK(sdkDiags)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Read current filter to get current items for diffing.
+	res, err := esClient.ML.GetFilters(esClient.ML.GetFilters.WithFilterID(compID.ResourceID), esClient.ML.GetFilters.WithContext(ctx))
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (edsavage) [reply internal/elasticsearch/ml/filter/update.go]: **Resolved** — Update parses `state.ID` with `CompositeIDFromStrFw` and uses `compID.ResourceID` for the ML API calls.

**Comment by edsavage** [comment]
Added acceptance coverage for **denied ML filter updates** when using `elasticsearch_connection` with a user that has no effective cluster privileges (`TestAccResourceMLFilterUpdateForbiddenReadOnlyUser` + testdata).

**Note:** that test expects the scoped Framework `elasticsearch_connection` credential behavior fixed in **#2882**. Merge **#2882** first, then rebase this branch onto current `main` before relying on the new ACC case in CI.

**Review by tobio — APPROVED** [review]
LGTM TY!
